### DISCUSSION TITLE
Update two related tables with one-to-many relation

### DISCUSSION THREAD

**Discussion by VladRafli** [prisma/prisma#18672] [Q&A] [answered]
Hello everyone

Im using NestJS in my project, when I tried to update a table with its children that has relation one-to-many, I got an error that says:

```
TypeError: t.map is not a function
    at tryInferArgs (E:\Work\Coding\Website\animal-adoption-nestjs\node_modules\@prisma\client\runtime\library.js:163:8028)
    at valueToArg (E:\Work\Coding\Website\animal-adoption-nestjs\node_modules\@prisma\client\runtime\library.js:163:5078)  
    at E:\Work\Coding\Website\animal-adoption-nestjs\node_modules\@prisma\client\runtime\library.js:163:8967
    at Array.reduce (<anonymous>)
    at objectToArgs (E:\Work\Coding\Website\animal-adoption-nestjs\node_modules\@prisma\client\runtime\library.js:163:8610)
    at tryInferArgs (E:\Work\Coding\Website\animal-adoption-nestjs\node_modules\@prisma\client\runtime\library.js:163:7154)
    at valueToArg (E:\Work\Coding\Website\animal-adoption-nestjs\node_modules\@prisma\client\runtime\library.js:163:5078)
    at E:\Work\Coding\Website\animal-adoption-nestjs\node_modules\@prisma\client\runtime\library.js:163:8967
    at Array.reduce (<anonymous>)
    at objectToArgs (E:\Work\Coding\Website\animal-adoption-nestjs\node_modules\@prisma\client\runtime\library.js:163:8610)
```

When trying to execute this query

```typescript
// PrismaService is service for accessing PrismaClient
await this.prismaService.animal.update({
      data: {
        animalPhoto: {
          updateMany: {
            data:
              animalPhoto !== undefined
                ? animalPhoto.map((val, idx) => ({
                    extension: uploadedFiles[idx].extension,
                    path: uploadedFiles[idx].path,
                    size: uploadedFiles[idx].size,
                    type: val.type,
                  }))
                : [],
            where: {
              animalId: id,
              deletedAt: null, // If Soft deleted
            },
          },
        },
        ...animalDto,
      },
      where: {
        id,
      },
    });
```

My prisma schema is:

```
model Animal {
  id           String        @id @unique @db.Char(40)
  userId       String        @db.Char(40)
  animalTypeId String        @db.Char(40)
  name         String        @db.VarChar(100)
  age          Int           @db.UnsignedTinyInt
  gender       AnimalGender
  breed        String        @db.VarChar(100)
  description  String?       @db.VarChar(255)
  createdAt    DateTime      @default(now()) @db.DateTime
  updatedAt    DateTime      @updatedAt @db.DateTime
  deletedAt    DateTime?     @db.DateTime
  user         User          @relation(fields: [userId], references: [id], onUpdate: Cascade, onDelete: NoAction)
  animalType   AnimalType    @relation(fields: [animalTypeId], references: [id], onUpdate: Cascade, onDelete: NoAction)
  animalPhoto  AnimalPhoto[]
}

model AnimalPhoto {
  id        String    @id @unique @db.Char(40)
  animalId  String    @db.Char(40)
  type      String    @db.VarChar(50)
  size      Int       @db.UnsignedInt
  extension String    @db.VarChar(100)
  path      String    @db.VarChar(255)
  createdAt DateTime  @default(now()) @db.DateTime
  updatedAt DateTime  @updatedAt @db.DateTime
  deletedAt DateTime? @db.DateTime
  animal    Animal    @relation(fields: [animalId], references: [id], onUpdate: Cascade, onDelete: NoAction)
}
```

I was wandering if the code is wrong, or it is a bug? Because the error stack came from Prisma Client, I have the workaround for this error by seperate the update query.

Thank you.

**Comment by nurul3101** [comment]
Hello @VladRafli 👋 

What is the value of the `animalPhoto` variable, could it be the case that animalPhoto is not undefined but is not an array as well?

**Reply by nurul3101** [reply]
Please let us know in case you are still facing this issue.

**Reply by VladRafli** [reply]
Im sorry for late reply.

If animalPhoto is not defined, it produce empty array, if defined, it produce array mapped by .map() function. I want to send the data when animalPhoto is present in my data, if not, i don't want to send animalPhoto as its undefined

**Reply by nurul3101** [reply]
I got it, but what I meant to say was that based on the error message that you are receiving `t.map is not a function`. To me, it looks like coming from this part `animalPhoto.map((val, idx) => ({`

I think value of `animalPhoto` would not be an array and something else due to which this part of the code throws an error. I would suggest you to log the value of `animalPhoto` and check.

**Reply by nurul3101** [reply]
If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by OlaQuteineh** [reply]
@VladRafli could you please share the work around or the fix you did as I'm facing the same issue :(

**Reply by VladRafli** [reply]
Im sorry for replying very lately.

> I got it, but what I meant to say was that based on the error message that you are receiving `t.map is not a function`. To me, it looks like coming from this part `animalPhoto.map((val, idx) => ({`
> 
> I think value of `animalPhoto` would not be an array and something else due to which this part of the code throws an error. I would suggest you to log the value of `animalPhoto` and check.

animalPhoto is input from user that cotains array of objects, and the input is validated that must be an array. This can be seen in here:

```ts
  // NestJS DTO
  // Decorator from library 'class-validator'
  @IsOptional()
  @IsArray()
  @ApiProperty()
  animalPhoto?: CreateAnimalPhotoDto[];
```

As the input is optional, `animalPhoto` may had a value of undefined, so I just make the logic inside the query to shorten the code. You can see this tenary operator

```ts
animalPhoto !== undefined ? 
         animalPhoto.map((val, idx) => ({
         extension: uploadedFiles[idx].extension,
         path: uploadedFiles[idx].path,
         size: uploadedFiles[idx].size,
         type: val.type,
    }))  // if true
    :
    [] // if false
```

So with this I can safely assume animalPhoto is not undefined in left operand in the tenary operator and do .map() operation. Maybe this would help to see where the problem is.

Thank you.

**Reply by VladRafli** [reply, marked as answer]
> @VladRafli could you please share the work around or the fix you did as I'm facing the same issue :(

You can take out the logic outside the query, and do some guard for the variable. And if the data is more than one, you can do some simple looping for updating the data one by one. (ps: It can be bad or uneffective due to updating data one by one, not in a bulk update)

```ts
if (animalPhoto !== undefined) {
    for (const photo of animalPhoto) {
        // ...code for update animal photos
    }
}

// ..code for update animal data
```

Hope this would help. 😊

**Reply by nurul3101** [reply]
Thanks for taking the time to share your solution 🙌
